### PR TITLE
Fix deferred credits

### PR DESCRIPTION
**Pull request by Eitu33**
* [x] document all added functions
* [ ] try in sandbox / simulation / labnet
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass
* [x] add logs allowing easy debugging in case the changes caused problems

### CODE REVIEW

**Comment by AurelienFT** [comment]
@Eitu33 Can you add some description on the changes and why ?

**Comment by Eitu33** [comment]
@AurelienFT @sydhds What this PR aims to fix:

1. Deferred credits changes were build to be absolute changes, meaning that the apply of speculative changes to the final deferred credits must override the values for any given existing entry. This is not what was done, the previous behaviour was accumulating the credits what would result in credits slashes never being applying because the change value was set to 0 and that 0 was later summed to the final value instead of replacing it.
2. The execution of deferred credits only credited the coins equivalent to the given address without setting the credit values to 0, what is even worst is that the only pruning operation we perform on credits removes only the credits that reached 0. The ram was increasing indefinitely because of that.
3. The previous deferred credits bootstrap didn't take into consideration that credits preceding the execution slot could change, with the denunciations coming they can, so the credits had to parallelly stream its changes in case some happened during the main credits bootstrap process.

File: massa-execution-worker/src/context.rs [hunk 1]
```diff
@@ -658,12 +658,27 @@ impl ExecutionContext {
...
-            if let Err(e) = self.transfer_coins(None, Some(addr), amount, false) {
+        let executed_credits = self.speculative_roll_state.get_deferred_credits(slot);
+
+        for (address, amount) in executed_credits {
+            // REVIEW NOTE: not sure which one is the best option here:
+            // * option 1: current one, requires to find the deferred_credits slot entry on every iteration even though the slot is always the same one
+            // * option 2: find the entry once but iterate twice because of the mutable reference incompatibility coming later from transfer_coins
+            // * option 3: get a copy of the slot credits first, modify the copy and replace the value in the deferred_credits after the iteration, but if transfer_coins fails no updates will be applied
+
+            self.speculative_roll_state
```
Comment (AurelienFT) [inline massa-execution-worker/src/context.rs:669]: Fetching in an hashmap is O(1) so there is not much problems fetching each slot I think.

**Review by AurelienFT — APPROVED** [review]

File: massa-execution-worker/src/context.rs (hunk 1 above)
Comment (Eitu33) [reply massa-execution-worker/src/context.rs:669]: Removing the review comment and merging

**Pull request merged**